### PR TITLE
tools/crimson: introduce crimson auto bench tool

### PR DESCRIPTION
**Pull request by aisakaki**
Write yaml file and run ./crimson_auto_bench.py to run multiple tests according to the yaml file and analyse results and draw graphics. 
Use --run to run the whole process, --bench to only run tests, --ana to organize data and draw graphics for exist bench results.
Use --repeat to decide the repeat times for every test.
All repeat results will be drawed as a point in the result graphics, and the mean value will be drawed as a mean line.   
By default, each tests configed in the yaml file are draw in dependent graphics, if you want to compare multiple tests in the config file, use --comp, and all tests will be drawed in one graphics. But be careful, when using --comp, you must write the same --x target config in config file.
You can set --x and multiple --y to decide what index the target graphics based on.
And also, all organized data will be stored in the graphics directory.
--help for more detail.
All failed tests(exceed retry limits of crimson_bench_tool) will be recorded at failure_log.txt
You can use tree root directory to see the structure
This tool require crimson_bench_tool.py at the same directory.

### CODE REVIEW

**Comment by aisakaki** [comment]
@cyx1231st  @liu-chunmei

File: tools/crimson/bench_config.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+fio_rbd_rand_write: 1
+client: 2 4
+block_size: 4096
+smp: 1 2
+time: 5
+tolerance_time: 5
+retry_limit: 2
+crimson: True
+store: bluestore
+iostat: 6
```
Comment (cyx1231st) [inline tools/crimson/bench_config.yaml:20]: Looks EOF is wrong.

File: tools/crimson/crimson_auto_bench.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        results = read_results(args.root)
+        res_path = f"{res_path_prefix}.{args.root}"
+        delete_and_create_at_local(res_path)
+        for y in args.y:
+            analysed_results = adjust_results(results, y)
+            draw(analysed_results, configs, args.x, y, res_path, args.comp)
+
+    if args.clean:
+        os.system("sudo rm -rf autobench.*")
+        os.system("sudo rm -rf graphic.autobench.*")
```
Comment (cyx1231st) [inline tools/crimson/crimson_auto_bench.py:301]: Looks EOF is wrong.

File: tools/crimson/crimson_bench_tool.py [hunk 3]
```diff
@@ -1304,9 +1315,10 @@ def check_failure(self):
                         help='how many osds')
     parser.add_argument('--log',
                         type=str,
-                        default = "log",
-                        help='directory to store logs, ./log by default. Will \
-                    store all tasks results and osd log and osd stdout')
+                        default = None,
+                        help='directory prefix to store logs, ./log_date by default.\
+                    will add _crimson/classic_backend_osd_poolsize to be log dir name. \
```
Comment (cyx1231st) [inline tools/crimson/crimson_bench_tool.py:1320]: Looks it is out-of-date?

File: tools/crimson/crimson_auto_bench.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+                        help= "remove all history graphic and bench results")
+
+    parser.add_argument('--config',
+                        type=str,
+                        default="bench_config.yaml",
+                        help="bench config yaml file path, bench_config.yaml by default")
+    parser.add_argument('--root',
+                        type=str,
+                        default=None,
+                        help="required if --ana")
```
Comment (cyx1231st) [inline tools/crimson/crimson_auto_bench.py:247]: Looks --root and --ana can be merged.

File: tools/crimson/crimson_auto_bench.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+    parser.add_argument('--repeat',
+                        type=int,
+                        default=1,
+                        help="repeat time for every tests, default to be 1")
+    parser.add_argument('--x',
+                        type=str,
+                        help="x axis of result graphics, required when --ana and --run")
```
Comment (cyx1231st) [inline tools/crimson/crimson_auto_bench.py:255]: Is there an example of what x could be, or where to identify the possible values of x ?

File: tools/crimson/crimson_auto_bench.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    parser.add_argument('--y',
+                        nargs='+',
+                        type=str,
+                        default=["IOPS"],
+                        help="y asix of result graphics, IOPS by default")
```
Comment (cyx1231st) [inline tools/crimson/crimson_auto_bench.py:260]: Looks y is only for the labeling purpose ?

Comment (aisakaki) [reply tools/crimson/crimson_auto_bench.py:260]: yes, all results from crimson bench tool will be recorded

File: tools/crimson/crimson_auto_bench.py (hunk 5 above)
Comment (aisakaki) [reply tools/crimson/crimson_auto_bench.py:255]: x can be smp, client, thread, osd_op_num_shards, ms_async_op_threads, etc.
all the parameter that can be multiple in the crimson bench tool can be x. Beacuse x is the main variable in the target graphics to draw x-y graphics

Comment (cyx1231st) [reply tools/crimson/crimson_auto_bench.py:255]: Let's elaborate them in the help message.

File: tools/crimson/crimson_auto_bench.py (hunk 6 above)
Comment (cyx1231st) [reply tools/crimson/crimson_auto_bench.py:260]: Then probably this is better:
"the label name of y asix of the result graphics, IOPS by default"

File: tools/crimson/crimson_bench_tool.py (hunk 3 above)
Comment (aisakaki) [reply tools/crimson/crimson_bench_tool.py:1320]: This means the default log prefix is "log_{date}" if there is no input.

File: tools/crimson/crimson_auto_bench.py (hunk 4 above)
Comment (aisakaki) [reply tools/crimson/crimson_auto_bench.py]: done

File: tools/crimson/crimson_auto_bench.py (hunk 2 above)
Comment (aisakaki) [reply tools/crimson/crimson_auto_bench.py:300]: fixed

**Review by cyx1231st — APPROVED** [review]

**Pull request merged**